serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/States.kt

```diff
@@ -78,44 +78,120 @@ class MatrixRooms(private val msession: MatrixSession, val message: String) : Ma
+                            edits.put(replaced_id,edit_msg)
+                            null
+                        } else {
+                            //No idea which event this edit is editing, just display fallback
+                            SharedUiMessage(it.sender, it.content.body, it.event_id, it.origin_server_ts)
+                        }
+                    } else {
+                        //This is a reply or normal text msg
+                        SharedUiMessage(it.sender, it.content.body, it.event_id, it.origin_server_ts, msg_content.relates_to?.in_reply_to?.event_id ?: "")
+                    }
+                }
+                is ImageRMEC -> generate_media_msg(msg_content.url, ::SharedUiImgMessage)
+                is AudioRMEC -> generate_media_msg(msg_content.url, ::SharedUiAudioMessage)
+                else -> SharedUiMessage(it.sender, "UNHANDLED EVENT!!! ${it.content.body}", it.event_id, it.origin_server_ts)
             }
         } else { null }
     }.filterNotNull()
-    fun sendMessage(msg: String, is_img: Boolean = false): MatrixState {
-        val sendFunc = if(is_img) { msession::sendImageMessage } else { msession::sendMessage }
-        when (val sendMessageResult = sendFunc(msg, room_id)) {
+    fun sendMessage(msg: String): MatrixState {
+        when (val sendMessageResult = msession.sendMessage(msg, room_id)) {
+            is Success -> { println("${sendMessageResult.value}") }
+            is Error -> { println("${sendMessageResult.message} - exception was ${sendMessageResult.cause}") }
+        }
+        return this
+    }
+    fun sendImageMessage(msg: String): MatrixState {
+        when (val sendMessageResult = msession.sendImageMessage(msg, room_id)) {
+            is Success -> { println("${sendMessageResult.value}") }
+            is Error -> { println("${sendMessageResult.message} - exception was ${sendMessageResult.cause}") }
+        }
+        return this
+    }
+    fun sendReply(msg: String, replied: String): MatrixState {
+        when (val sendMessageResult = msession.sendMessage(msg, room_id, replied)) {
             is Success -> { println("${sendMessageResult.value}") }
             is Error -> { println("${sendMessageResult.message} - exception was ${sendMessageResult.cause}") }
         }
         return this
     }
+    fun sendEdit(msg: String, edited_id: String): MatrixState {
+        when (val sendMessageResult = msession.sendEdit(msg, room_id, edited_id)) {
+            is Success -> { println("${sendMessageResult.value}") }
+            is Error -> { println("${sendMessageResult.message} - exception was ${sendMessageResult.cause}") }
+        }
+        return this
+    }
+    fun getEventSrc(msg_id: String): String {
+        for(event in msession.getRoomEvents(room_id)) {
+            if (event as? RoomMessageEvent != null) {
+                if(event.event_id == msg_id) {
+                    return event.raw_self.toString()
```

Comment:
Hehehe, there has to be a way to get the pretty printed JSON, though I admit I didn't find it with some quick searching...

Reply:
I was so frustrated trying to find an easy way. The only way I found readily available was string -> obj -> specialized format object from serialization lib -> back to string and that seemed way too overkill for the debug feature.